File: tbb/chain/seq/randmat.cc
/* randmat: random number generator
 *
 * input:
 *   nrows, ncols: number of rows and columns
 *   s: random number generation seed
 *
 * output:
 *   matrix: random nrows x ncols integer matrix
 */
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cstring>

unsigned char randmat_matrix[30000][30000];

void randmat(int nrows, int ncols, unsigned int seed) {
  const int LCG_A = 1664525, LCG_C = 1013904223;
  for (int i = 0; i < nrows; i++) {
    for (int j = 0; j < ncols; j++) {
      randmat_matrix[i][j] = seed = (LCG_A * seed + LCG_C) % 100;
    }
  }
}

File: erlang/chain/expertseq/randmat.erl
%
% randmat: random number generation
%
% input:
%   nrows, ncols: the number of rows and columns
%   s: the seed
%
% output:
%   matrix: a nrows x ncols integer matrix
%

-module(randmat).
-export([randmat/3]).
-define(INT_MAX,4294967296).
-define(RAND_MAX,100).
-define(LCG_A,1664525).
-define(LCG_C,1013904223).

randvet(0, _) -> [];
randvet(Ncols, S) ->
  NewS = (?LCG_A * S + ?LCG_C) rem ?INT_MAX,
  [NewS rem ?RAND_MAX | randvet(Ncols - 1, NewS)].

randvet_acc (0, _, Acc) ->
    lists:reverse (Acc);
randvet_acc (Ncols, S, Acc) ->
    NewS = (?LCG_A * S + ?LCG_C) rem ?INT_MAX,
    Val = NewS rem ?RAND_MAX,
    randvet_acc (Ncols - 1, NewS, [Val] ++ Acc).


randmat(Nrows, Ncols, S) ->
    randmat_acc (Nrows, Ncols, S, []).

randmat_acc (0, _, _, Acc) -> lists:reverse (Acc);
randmat_acc (Nrows, Ncols, S, Acc) -> 
    randmat_acc (Nrows - 1, Ncols, S+1, [randvet (Ncols, S)] ++ Acc).
